lib/Connection.php

```diff
@@ -335,6 +344,18 @@ public function query($sql, &$values=array())
 				throw new DatabaseException($this);
 		} catch (PDOException $e) {
 			throw new DatabaseException($e);
+		} finally {
```

Comment:
What do you think about this approach, to reduce the amount of code in the catch blocks:

```php
	public function query($sql, &$values=array())
	{
		if ($this->logging)
		{
			$start = microtime(true);
			$this->logger->log($sql);
			if ( $values ) $this->logger->log($values);
		}

		$log_execution_time = function($logger, $start) {
			$time_elapsed_ms = (microtime(true) - $start) * 1000;
			$this->total_elapsed_query_time += $time_elapsed_ms;
			$logger->log(
				sprintf(
					'Query took %dms to execute. (%dms total)',
					$time_elapsed_ms,
					$this->total_elapsed_query_time
				)
			);
		};

		$this->last_query = $sql;

		try {
			if (!($sth = $this->connection->prepare($sql)))
				throw new DatabaseException($this);
		} catch (PDOException $e) {
			throw new DatabaseException($this);
		}

		$sth->setFetchMode(PDO::FETCH_ASSOC);

		try {
			if (!$sth->execute($values))
				throw new DatabaseException($this);

			if (isset($start) && $this->logging) {
				$log_execution_time($this->logger, $start);
			}
		} catch (PDOException $e) {
			if (isset($start) && $this->logging) {
				$log_execution_time($this->logger, $start);
			}

			throw new DatabaseException($e);
		} catch (Exception $e) {
			if (isset($start) && $this->logging) {
				$log_execution_time($this->logger, $start);
			}

			return $e;
		}
		return $sth;
	}
```